dual_net.py

```diff
@@ -321,58 +193,133 @@ def my_res_layer(inputs, train_mode):
-        'value_cost': value_cost,
-        'l2_cost': l2_cost,
-        'combined_cost': combined_cost,
-        'policy_entropy': policy_entropy,
-        'global_step': global_step,
-        'train_op': train_op,
-    }
-
 
-def logging_ops():
-    return tf.summary.merge([
-        tf.summary.histogram(weight_var.name, weight_var)
-        for weight_var in tf.trainable_variables()],
-        name="weight_summaries")
+    metric_ops = {
+        'accuracy': tf.metrics.accuracy(labels=labels['pi_tensor'],
+                                        predictions=policy_output,
+                                        name='accuracy_op'),
+        'policy_cost': tf.metrics.mean(policy_cost),
+        'value_cost': tf.metrics.mean(value_cost),
+        'l2_cost': tf.metrics.mean(l2_cost),
+        'policy_entropy': tf.metrics.mean(policy_entropy),
+        'combined_cost': tf.metrics.mean(combined_cost),
+    }
+    # Register metrics as tf.summary to make them show up during training
+    for metric_name, metric_op in metric_ops.items():
+        tf.summary.scalar(metric_name, metric_op[1])
+
+    return tf.estimator.EstimatorSpec(
+        mode=mode,
+        predictions={
+            'policy_output': policy_output,
+            'value_output': value_output,
+        },
+        loss=combined_cost,
+        train_op=train_op,
+        eval_metric_ops=metric_ops,
+        )
+
+
+def get_estimator(working_dir, **hparams):
+    hparams = get_default_hyperparams(**hparams)
+    return tf.estimator.Estimator(
+        model_fn,
+        model_dir=working_dir,
+        params=hparams)
+
+
+def bootstrap(working_dir, **hparams):
+    """Initialize a tf.Estimator run with random initial weights.
+
+    Args:
+        working_dir: The directory where tf.estimator will drop logs,
+            checkpoints, and so on
+        hparams: hyperparams of the model.
+    """
+    hparams = get_default_hyperparams(**hparams)
+    # a bit hacky - forge an initial checkpoint with the name that subsequent
+    # Estimator runs will expect to find.
+    #
+    # Estimator will do this automatically when you call train(), but calling
+    # train() requires data, and I didn't feel like creating training data in
+    # order to run the full train pipeline for 1 step.
+    estimator_initial_checkpoint_name = 'model.ckpt-1'
+    save_file = os.path.join(working_dir, estimator_initial_checkpoint_name)
+    sess = tf.Session(graph=tf.Graph())
+    with sess.graph.as_default():
+        features, labels = get_inference_input()
+        model_fn(features, labels, tf.estimator.ModeKeys.PREDICT, hparams, None)
+        sess.run(tf.global_variables_initializer())
+        tf.train.Saver().save(sess, save_file)
+
+
+def export_model(working_dir, model_path):
+    """Take the latest checkpoint and export it to model_path for selfplay.
+
+    Assumes that all relevant model files are prefixed by the same name.
+    (For example, foo.index, foo.meta and foo.data-00000-of-00001).
```

Comment:
maybe later we can export & optimize these graphs for inference as a single file.

Reply:
Hm. Eventually, yes. The default estimator exporter creates a function signature that isn't based on feed_dict, but rather, hardwired to accept serialized tf.Example objects. Which is to say, I think the Estimator API expects deserialization to happen in the model_fn rather than the input_fn. But I put the deserialization code in the input_fn so that we could use feed_dict as an alternate input pathway into model_fn. Some wrangling will be necessary later on, I think, if we want to make C++ inference work.